Add automatic checking for engine updates

KoBeWi opened the pull request:
Closes https://github.com/godotengine/godot-proposals/issues/1132

![](https://private-user-images.githubusercontent.com/2223172/307879982-cf14c0cb-c1a4-4201-8750-ebbd041e47b2.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDg5Nzk4NzcsIm5iZiI6MTcwODk3OTU3NywicGF0aCI6Ii8yMjIzMTcyLzMwNzg3OTk4Mi1jZjE0YzBjYi1jMWE0LTQyMDEtODc1MC1lYmJkMDQxZTQ3YjIucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDIyNiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAyMjZUMjAzMjU3WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MWY4MzUzNjBjN2ZhNDliMTEzNjM0ZWZiOGJkMzM1M2Q3ZTRmMWU4M2RjMDc5MDZmYjljYzAwZWMwYWIzNGNiOSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.e2hSh3iFavpCD08GDCh92yAJuK6plSxM2nHddNAGLfI)
The PR adds update checks in a form of a label on the bottom of Project Manager and editor. See https://github.com/godotengine/godot/pull/75916#issuecomment-1964669768 for more images.

The update checker will check for newer version and tell you when one is available. Clicking the button will open the download page for this specific version. The checker can check updates in 3 modes: "Newest Unstable", which checks for newest official version, "Newest Stable", which checks for newest stable version and "Newest Patch", which checks for the newest patch of the current minor version. "Newest Stable" is the default for stable versions, "Newest Unstable" is default for preview versions.

Note that updates are disabled by default due to network mode being offline by default. There is also a setting to disable updates separately from network mode.

For version checking I'm parsing the [version YAML](https://raw.githubusercontent.com/godotengine/godot-website/master/_data/versions.yml) from Godot website. The editor will make many assumptions about structure of the file (such as it's sorted from newest to oldest version), so hopefully it won't change 😅

You can test the PR by modifying `version.py` file. Compiling with `allow_automatic_update=false` scons option will disable the dialog completely.

KoBeWi committed:
Add automatic checking for engine updates

Calinou commented:
Remember to add a SCons option to allow disabling update checks at build-time, as the Flatpak (and other Linux distribution packages) must not display in-app update notifications by design.

KoBeWi commented:
> This is looking pretty cool, but i think the check for updates should be only in the project manager and not in the editor as it can be distracting as users could enable it by accident there.

The popup appears on first launch and then only when there is an update available. Other than that the dialog will check for new versions in the background and you won't see it very often.
I added it in the editor too, because it's possible to open the project directly, and while probably rare, some people don't use Project Manager very much or at all.

> Also maybe an option to replace the current version with the newer one so you don't have to uninstall a old version you don't want to use for a more stable one.

This is not that simple, because you can't replace an executable that is already running (at least on Windows). The implementation would be a bit involved. For now only version checking will be available; actual update can be added later.

KoBeWi commented:
The dialog that appears is the same both for Project Manager and editor. Before making any HTTP requests it will ask you if you want to enable it (unlike AssetLib btw). The setting is disabled by default.

capnm commented:
> The dialog that appears is the same both for Project Manager and editor. Before making any HTTP requests it will ask you if you want to enable it (unlike AssetLib btw). The setting is disabled by default.

Why not just integrate the (annoying) dialog into the manager? Then you don't have to bother about dev builds.

RedworkDE commented on editor/engine_update_dialog.cpp:
```diff
@@ -0,0 +1,219 @@
+	}
+
+	if (max_version > current_version) {
+		if (max_version_string == EditorSettings::get_singleton()->get_meta("auto_updates_skip_version", "").operator String()) {
```
This mean that if I skip the update to a new minor version (maybe it breaks something in my project) I will not be notified about new patch version for my current minor (that I may want to install), just for patches for the new minor version (that I don't care about)

RedworkDE commented on editor/engine_update_dialog.cpp:
```diff
@@ -0,0 +1,219 @@
+	bool is_stable = false;
+	for (const String &line : p_content) {
+		print_line(line);
+		if (line.contains("stable")) {
```
If I am currently on a pre-release version I probably do want to update to a newer pre-release, not just the stable version.

RedworkDE reviewed (commented):
Not sure how much traffic this causes, but I would probably work on a document just listing the interesting versions to minimize it and possibly simplify this PR a bit.

xix-xeaon commented:
I like being notified about updates. I do not like to be interrupted by dialogs. I think a passive icon should show up in the top right corner (or on the bottom, next to the display of the current version) when there's a new update.

jordanlis commented:
> I like being notified about updates. I do not like to be interrupted by dialogs. I think a passive icon should show up in the top right corner (or on the bottom, next to the display of the current version) when there's a new update.

I agree with Xix, even if I realize that this pop-up has been developed. 
I imagined something way less distracting

![image](https://github.com/godotengine/godot/assets/68544949/0ca75810-bb98-4a33-a51f-19afbc38fcf3)

The mockup is not perfect, but the idea is here. 

Anyway, it's still a great feature that is really good, even in a pop-up.

KoBeWi commented:
Ok I have an idea how for alternative implementation (which is similar to above comments). This is what would users see by default:
![image](https://github.com/godotengine/godot/assets/2223172/b9fa4264-2ff6-4d76-adf1-58243828931a)
The `[ Auto-update is disabled ]` (style TBD) would be a clickable button, which would open a dialog similar to the currently implemented popup. Until the user enables auto-updates, Godot will not do version checks of any kind.

If the update is enabled, the text will change depending on version status:
- `[ No update available ]`
- `[ New stable version available: VERSION ]`
- `[ New unstable version available: VERSION ]`
- `[ Failed to check for updates ]`

The stable/unstable text could have different tints. Failed text would allow retrying the check.

I'm not updating the PR yet, because the future of this feature is actually uncertain right now.

KoBeWi commented:
I pushed a WIP update. It does more or less what I described in the above comment:
![image](https://github.com/godotengine/godot/assets/2223172/cf14c0cb-c1a4-4201-8750-ebbd041e47b2)
Clicking it opens the version page in download archive. I also changed how version is determined using the [YAML file from Godot website](https://raw.githubusercontent.com/godotengine/godot-website/master/_data/versions.yml) (my code makes many assumptions about it's format, but that's *probably ok*).

If the user is in offline mode, updates are disabled:
![image](https://github.com/godotengine/godot/assets/2223172/0df4263e-7af6-491a-b9ef-9f098323d6c7)

If there is no update:
![image](https://github.com/godotengine/godot/assets/2223172/4b1eabad-e7ee-42a3-8084-0c9d84d228ce)

Also this mostly works with official versions. In custom builds it can fail if it can't determine newer version:
![image](https://github.com/godotengine/godot/assets/2223172/ae2362c4-f162-4b03-b42a-a83f2fc3a725)

I still have a cleanup to do and I plan I plan to add an editor setting. The setting would have 4 values: "Disable Update Checks", "Check Newest Unstable", "Check Newest Stable" and "Check Newest Patch".

Also the update checker is only in project manager right now. I will probably add it to EditorNode next to the version label, but it'st going to behave a bit differently to not take space.

Calinou commented:
The new button looks good, but I'd make the "Current version up-to-date" text gray like the offline mode text so it doesn't take as much attention from the rest of the dialog. Only "update available" and "failed to check for updates" should really stand out from the rest.

KoBeWi commented:
Ok I changed the up-to-date text:
![image](https://github.com/godotengine/godot/assets/2223172/587aa772-5230-455f-830c-d8264cb880aa)

Also added the update label to the editor:
![image](https://github.com/godotengine/godot/assets/2223172/70e2890f-fabc-4778-a30f-d3fe9d54bdfc)
However it will only be visible if it's currently checking for updates or an update is available. If there is an error or no new version, the label will be hidden to save space. In project manager it's always visible.

I also added the "Newest Stable" and "Newest Patch" update modes.
![image](https://github.com/godotengine/godot/assets/2223172/42156d55-4212-4b9a-96aa-6b18398b66f6)
![image](https://github.com/godotengine/godot/assets/2223172/c61a0179-e18d-4d79-a5a5-90f78180f4e9)
Newest Stable is the default.

With this the PR should be ready to go.

Calinou reviewed (commented):
Tested locally, it works as expected.

However, when running a self-compiled development build, you'll always see this at the bottom of the project manager:

![image](https://github.com/godotengine/godot/assets/180032/f06d5cd6-b838-49e1-b795-bd3dae974d98)

The red color can make people think it's an error, leading to false bug reports as update checking for these builds isn't supported by design.

To handle this particular situation (unnumbered `dev`/`beta`/`rc` builds), I suggest displaying a message in gray of the form:

> Running a development build.

KoBeWi commented:
btw I just had an interesting idea. Say we merge this for some dev version. Since "Newest Stable" is the default setting, when a beta appears, users will still see that there is no update, which might lead to bug reports. What if the default was different for stable versions and unstable versions? This way if you have unstable version, you will get updates until stable, then in stable you will only get info about next stable, unless you again go to unstable build. Does that make sense?

Calinou commented:
> What if the default was different for stable versions and unstable versions? This way if you have unstable version, you will get updates until stable, then in stable you will only get info about next stable, unless you again go to unstable build. Does that make sense?

Stable versions should only inform you about other stable versions by default, while unstable versions should inform you about unstable versions as well.

timothyqiu commented:
By the way, I think the network request should also utilize `network/http_proxy/{host,port}` editor settings.

akien-mga commented on SConstruct:
```diff
@@ -467,6 +468,9 @@ if methods.get_cmdline_bool("fast_unsafe", env_base.dev_build):
 if env_base["use_precise_math_checks"]:
     env_base.Append(CPPDEFINES=["PRECISE_MATH_CHECKS"])
 
+if env_base["allow_automatic_updates"]:
+    env_base.Append(CPPDEFINES=["ALLOW_AUTO_UPDATE"])
```
Both the option and the define sound like the engine will auto-update, when it just does update checks. So I would rename like this:
```suggestion
if env_base["engine_update_check"]:
    env_base.Append(CPPDEFINES=["ENGINE_UPDATE_CHECK_ENABLED"])
```

akien-mga commented on doc/classes/EditorSettings.xml:
```diff
@@ -636,6 +636,14 @@
+		<member name="interface/editor/engine_version_update_mode" type="int" setter="" getter="">
+			Specifies how the engine should check for updates.
+			- [b]Disable Update Checks[/b] will block the engine from checking updates (see also [member network/connection/network_mode]).
+			- [b]Check Newest Unstable[/b] (default for preview versions) will check for newest available official version.
```
```suggestion
			- [b]Check Newest Preview[/b] (default for preview versions) will check for the newest available development snapshot.
```

akien-mga commented on doc/classes/EditorSettings.xml:
```diff
@@ -636,6 +636,14 @@
+			Specifies how the engine should check for updates.
+			- [b]Disable Update Checks[/b] will block the engine from checking updates (see also [member network/connection/network_mode]).
+			- [b]Check Newest Unstable[/b] (default for preview versions) will check for newest available official version.
+			- [b]Check Newest Stable[/b] (default for stable versions) will check for newest available stable version.
```
```suggestion
			- [b]Check Newest Stable[/b] (default for stable versions) will check for the newest available stable version.
```

akien-mga commented on doc/classes/EditorSettings.xml:
```diff
@@ -636,6 +636,14 @@
+			- [b]Disable Update Checks[/b] will block the engine from checking updates (see also [member network/connection/network_mode]).
+			- [b]Check Newest Unstable[/b] (default for preview versions) will check for newest available official version.
+			- [b]Check Newest Stable[/b] (default for stable versions) will check for newest available stable version.
+			- [b]Check Newest Patch[/b] will check for newest available stable version, but only within the same minor version. E.g. if your version is [code]4.3[/code], you will be notified about [code]4.3.1[/code], but not [code]4.4[/code].
```
```suggestion
			- [b]Check Newest Patch[/b] will check for the latest available stable version, but only within the same minor version. E.g. if your version is [code]4.3.stable[/code], you will be notified about [code]4.3.1.stable[/code], but not [code]4.4.stable[/code].
```

akien-mga commented on doc/classes/EditorSettings.xml:
```diff
@@ -636,6 +636,14 @@
+			- [b]Check Newest Unstable[/b] (default for preview versions) will check for newest available official version.
+			- [b]Check Newest Stable[/b] (default for stable versions) will check for newest available stable version.
+			- [b]Check Newest Patch[/b] will check for newest available stable version, but only within the same minor version. E.g. if your version is [code]4.3[/code], you will be notified about [code]4.3.1[/code], but not [code]4.4[/code].
+			All update modes will ignore builds with different major version (e.g. Godot 4 -&gt; Godot 5).
```
```suggestion
			All update modes will ignore builds with different major versions (e.g. Godot 4 -&gt; Godot 5).
```

akien-mga commented on doc/classes/EditorSettings.xml:
```diff
@@ -861,7 +869,7 @@
 		</member>
 		<member name="network/connection/network_mode" type="int" setter="" getter="">
-			Determines whether online features are enabled in the editor, such as the Asset Library. Setting this property to "Offline" is recommended to limit editor's internet activity, especially if privacy is a concern.
+			Determines whether online features are enabled in the editor, such as the Asset Library or update checks. Setting this property to "Offline" is recommended to limit editor's internet activity, especially if privacy is a concern.
```
I know it's pre-existing, but I don't think we should say setting this to "Offline" is _recommended_.
```suggestion
			Determines whether online features are enabled in the editor, such as the Asset Library or update checks. Disabling these online features helps alleviate privacy concerns by preventing the editor from making HTTP requests to the Godot website, GitHub, or third-party platforms hosting assets from the Asset Library.
```

akien-mga commented on editor/editor_settings.cpp:
```diff
@@ -413,6 +414,14 @@ void EditorSettings::_load_defaults(Ref<ConfigFile> p_extra_config) {
+		if (String(VERSION_STATUS) == String("stable")) {
+			default_update_mode = EngineUpdateLabel::UpdateMode::NEWEST_STABLE;
+		}
+		EDITOR_SETTING(Variant::INT, PROPERTY_HINT_ENUM, "interface/editor/engine_version_update_mode", int(default_update_mode), "Disable Update Checks,Check Newest Unstable,Check Newest Stable,Check Newest Patch"); // Uses EngineUpdateLabel::UpdateMode.
```
```suggestion
		EDITOR_SETTING(Variant::INT, PROPERTY_HINT_ENUM, "interface/editor/engine_version_update_mode", int(default_update_mode), "Disable Update Checks,Check Newest Preview,Check Newest Stable,Check Newest Patch"); // Uses EngineUpdateLabel::UpdateMode.
```

KoBeWi commented:
> (Or retrigger the dialog that asks about it maybe? Or was it a full screen thingy in empty assetlib?).

That dialog is removed.

akien-mga commented on editor/gui/editor_bottom_panel.cpp:
```diff
@@ -247,6 +248,13 @@ EditorBottomPanel::EditorBottomPanel() {
 	button_hbox->set_h_size_flags(Control::SIZE_EXPAND_FILL);
 	bottom_hbox->add_child(button_hbox);
 
+#ifdef ALLOW_AUTO_UPDATE
+	EngineUpdateLabel *eul = memnew(EngineUpdateLabel);
+	eul->set_v_size_flags(Control::SIZE_SHRINK_CENTER);
+	eul->enable_compact_mode();
+	bottom_hbox->add_child(eul);
+#endif
```
I wonder if we even need it in the editor. I think having it in the project manager might be good enough.
If a user chooses to stay on an older version for one reason or another, but still wants to keep track of the latest, it makes sense to me to see the update message in the PM but not to have it nag you constantly during development in the editor.

We're also constrained in width in the editor bottom panel, this message being constantly there will quickly become a problem on smaller screens.

KoBeWi replied:
This option is for users who don't use Project Manager. But I guess not using it at all is not that common 🤔

akien-mga replied:
Yeah I thought about that too. It's true alternative project managers like `godots` are rising in popularity and likely imply skipping the default project manager. But those alternative PMs also do their own update checking logic, so it might not be much of a problem.

KoBeWi replied:
What I mean is launching project directly, like directly from the folder or using e.g. VS Code. Personally I open projects with VS Code most of the time.

akien-mga reviewed (commented):
Quick code and doc review.

akien-mga reviewed (approved):
Looks great to me!

akien-mga commented:
Thanks!

akien-mga commented on editor/engine_update_label.cpp:
```diff
@@ -0,0 +1,344 @@
+}
+
+void EngineUpdateLabel::_http_request_completed(int p_result, int p_response_code, const PackedStringArray &p_headers, const PackedByteArray &p_body) {
+	if (p_result != OK) {
+		_set_status(UpdateStatus::ERROR);
+		_set_message(vformat(TTR("Failed to check for updates. Error: %d."), p_result), theme_cache.error_color);
+		return;
+	}
+
+	if (p_response_code != 200) {
+		_set_status(UpdateStatus::ERROR);
+		_set_message(vformat(TTR("Failed to check for updates. Response code: %d."), p_response_code), theme_cache.error_color);
+		return;
+	}
```
Just thought about this, but I wonder if we should future proof in case we ever refactor the website and no longer have https://raw.githubusercontent.com/godotengine/godot-website/master/_data/versions.yml

Maybe we should just print verbose these messages, and not show anything in the PM when requests fail?

So people can keep using e.g. Godot 4.3-stable without being nagged if in a year the website has been ported to a new framework and this file no longer exists.

Otherwise, we need to make sure the file we use for these checks will always be online (but that's hard to future proof).

WDYT @Calinou @coppolaemilio?

fire replied:
This is a valid concern because we might lose control of githubusercontent.com. However it can be debated we can also lose control of godotengine.org.

akien-mga replied:
Yeah I was thinking the same.

We currently add .json files to https://godotengine.org/mirrorlist/ (e.g. https://godotengine.org/mirrorlist/4.2.1.stable.json) for each release, but it's not the most convenient to parse.

It should be fairly easy to add a build step for the website to generate e.g. https://godotengine.org/mirrorlist/versions.json that would include all versions with their respective blog post.

coppolaemilio replied:
Yes, we shouldn't use this file. I would go with our website. We can serve a page like versions.yml in a domain that we control so that if we change servers or things like that we still have the same response from the same address. 

This could be `versions.godotengine.org/yml` or `godotengine.org/versions.yml`. Anything like that will be better.

I can make the changes needed for getting that address up.

akien-mga replied:
BTW when we implement that, the description I amended in https://github.com/godotengine/godot/pull/75916/files#diff-52f0c165415d19abd2d5586c8a168169a8d39716be86688a9117125409b421f0R878 should be changed to remove explicit mention of GitHub.
